modules/heatpump.py

```diff
@@ -0,0 +1,133 @@
+
+        Raise:
+            ValueError: If outside temperature isn't range.
+
+        Return:
+            cop: Calculated COP based on temperature
+        """
+        # TODO: Support for other temperature units (e.g Fahrenheit, Kelvin)
+        # Check for sensible temperature values
+        if self.__check_outside_temperature_range__(outside_temperature_celsius):
+            cop = self.COP_BASE + (outside_temperature_celsius * self.COP_COEFFICIENT)
+            return max(cop, 1)
+        else:
+            err_msg = f"Outside temperature '{outside_temperature_celsius}' not in range (min: -100 Celsius, max: 100 Celsius) "
+            self.log.error(err_msg)
+            raise ValueError(err_msg)
+
+    def calculate_heating_output(self, outside_temperature_celsius: float) -> float:
+        """Calculate the heating output in Watts based on outside temperature in degree Celsius.
+        Temperature range must be between -100 and 100 degree Celsius.
+
+        Args:
+            outside_temperature_celsius: Outside temperature in degree Celsius
+
+        Raises:
+            ValueError: Raised if outside temperature isn't in described range.
+
+        Returns:
+            heating output: Calculated heating output in Watts.
+        """
+        if self.__check_outside_temperature_range__(outside_temperature_celsius):
+            heat_output = (
+                (
+                    self.BASE_HEATPOWER
+                    + outside_temperature_celsius * self.TEMPERATURE_COEFFICIENT
+                )
+                * 1000
+            ) / 24.0
+            return min(self.max_heat_output, heat_output)
+        else:
+            err_msg = f"Outside temperature '{outside_temperature_celsius}' not in range (min: -100 Celsius, max: 100 Celsius) "
+            self.log.error(err_msg)
+            raise ValueError(err_msg)
+
+    def calculate_heat_power(self, outside_temperature_celsius: float) -> float:
+        """Calculate electrical power based on outside temperature (degree Celsius).
+
+        Args:
+            outside_temperature_celsius: Temperature in range -100 to 100 degree Celsius.
+
+        Raises:
+            ValueError: Raised if temperature isn't in described range
+
+        Returns:
+            power: Calculated electrical power in Watt.
+        """
+        if self.__check_outside_temperature_range__(outside_temperature_celsius):
+            return (
+                1164
+                - 77.8 * outside_temperature_celsius
+                + 1.62 * outside_temperature_celsius**2.0
+            )
+        else:
+            err_msg = f"Outside temperature '{outside_temperature_celsius}' not in range (min: -100 Celsius, max: 100 Celsius) "
+            self.log.error(err_msg)
+            raise ValueError(err_msg)
+
+    def simulate_24h(self, temperatures: List[float]) -> List[float]:
+        """Simulate power data for 24 hours based on provided temperatures."""
+        power_data = List[float]
```

Comment:
Rule of thumb: Use invariant types (`Sequence`, `Collection`, `Iterable`, `Mapping`, ...) for parameters and covariant types (`List`, `Tuple`, `Dict`, ...) for return values.
This keeps the code flexible while improving the utility of type information.

In this example: `.simulate_24h(temperatures=(1,2,3,...,24))` works with the implementation, but wouldn't be allowed if `List[float]` was required for the input.

```suggestion
    def simulate_24h(self, temperatures: Sequence[float]) -> List[float]:
        """Simulate power data for 24 hours based on provided temperatures."""
        power_data: List[float] = []
```